Move "parse a list of dimensions" implementation off of Length

weinig opened the pull request:
#### ca4ac329e42dc2c254933cbc0b8e0f24ea0f7f88
<pre>
Move &quot;parse a list of dimensions&quot; implementation off of Length
<a href="https://bugs.webkit.org/show_bug.cgi?id=296909">https://bugs.webkit.org/show_bug.cgi?id=296909</a>

Reviewed by Darin Adler.

As part of the effort to get code off the overly general Length class,
this moves the HTMLFrameSetElement rows/cols parsing implementation
off of Length and onto a bespoke type using the standardized parsing.

The new behavior is tested via new unit tests WebCoreHTMLParserIdioms.parseHTMLDimensionsList.

As a next step, aligning the layout algorithm with the standard would
allow un-skipping the WPT frameset tests in html/rendering/non-replaced-elements/the-frameset-and-frame-elements/.

* Source/WTF/wtf/text/StringParsingBuffer.h:
* Source/WebCore/html/HTMLFrameSetElement.cpp:
* Source/WebCore/html/HTMLFrameSetElement.h:
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
* Source/WebCore/html/parser/HTMLParserIdioms.h:
* Source/WebCore/platform/Length.cpp:
* Source/WebCore/platform/Length.h:
* Source/WebCore/rendering/RenderFrameSet.cpp:
* Source/WebCore/rendering/RenderFrameSet.h:
* Tools/TestWebKitAPI/Tests/WebCore/HTMLParserIdioms.cpp:

Canonical link: <a href="https://commits.webkit.org/298436@main">https://commits.webkit.org/298436@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eca16f1a9fb8b2471f57be2d316e49ec293515dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115452 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35143 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25645 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121537 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66024 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117341 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35804 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43714 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87733 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118400 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28558 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103633 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68125 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27717 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21755 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65193 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107649 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97948 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124701 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113988 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42387 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31757 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96501 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42754 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99822 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96287 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24502 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41516 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19378 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38285 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42267 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47827 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/139471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41770 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/139471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45096 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43486 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -522,4 +525,157 @@ std::optional<HTMLDimension> parseHTMLMultiLength(StringView multiLengthString)
+            // 5.7.3.3. Increment value by fraction.
+
+            if (length > 0)
+                value += (static_cast<double>(fraction) / std::pow(10, static_cast<double>(length)));
```
Why the static_cast? I think this would do the exact same operation:
```suggestion
                value += fraction / std::pow(10.0, length);
```

darinadler commented on Source/WebCore/rendering/RenderFrameSet.cpp:
```diff
@@ -188,26 +189,26 @@ void RenderFrameSet::layOutAxis(GridAxis& axis, std::span<const Length> grid, in
-        if (grid[i].isFixed()) {
-            gridLayout[i] = std::max(grid[i].intValue(), 0);
+        if (grid[i].unit == HTMLDimensionsListValue::Unit::Absolute) {
+            gridLayout[i] = std::max(static_cast<int>(grid[i].number), 0);
```
Will this work?
```suggestion
            gridLayout[i] = std::max<int>(grid[i].number, 0);
```
Is the version with a cast better? Is it best to truncate rather than rounding?

weinig replied:
This whole function is going to be replaced, so, since truncation preserves the existing behavior, I'm going to keep it and re-evaluate in the replacement.

darinadler commented on Source/WebCore/rendering/RenderFrameSet.cpp:
```diff
@@ -188,26 +189,26 @@ void RenderFrameSet::layOutAxis(GridAxis& axis, std::span<const Length> grid, in
-        if (grid[i].isPercentOrCalculated()) {
-            gridLayout[i] = std::max(intValueForLength(grid[i], availableLen), 0);
+        if (grid[i].unit == HTMLDimensionsListValue::Unit::Percentage) {
+            gridLayout[i] = std::max(static_cast<int>(availableLen * grid[i].number / 100.0), 0);
```
Will the below work? Same thought for the 3 other cases below. I suppose it's nice to be able to see explicit truncation so we can ask whether we want flooring, rounding, ceiling or something else, but also would prefer not to use a cast for a conversion that works without one.
```suggestion
            gridLayout[i] = std::max<int>(availableLen * grid[i].number / 100.0, 0);
```

darinadler replied:
will this work?
```suggestion
            gridLayout[i] = std::max<int>(availableLen * grid[i].number / 100.0, 0);
```

weinig committed:
Move "parse a list of dimensions" implementation off of Length
https://bugs.webkit.org/show_bug.cgi?id=296909

Reviewed by Darin Adler.

As part of the effort to get code off the overly general Length class,
this moves the HTMLFrameSetElement rows/cols parsing implementation
off of Length and onto a bespoke type using the standardized parsing.

The new behavior is tested via new unit tests WebCoreHTMLParserIdioms.parseHTMLDimensionsList.

As a next step, aligning the layout algorithm with the standard would
allow un-skipping the WPT frameset tests in html/rendering/non-replaced-elements/the-frameset-and-frame-elements/.

* Source/WTF/wtf/text/StringParsingBuffer.h:
* Source/WebCore/html/HTMLFrameSetElement.cpp:
* Source/WebCore/html/HTMLFrameSetElement.h:
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
* Source/WebCore/html/parser/HTMLParserIdioms.h:
* Source/WebCore/platform/Length.cpp:
* Source/WebCore/platform/Length.h:
* Source/WebCore/rendering/RenderFrameSet.cpp:
* Source/WebCore/rendering/RenderFrameSet.h:
* Tools/TestWebKitAPI/Tests/WebCore/HTMLParserIdioms.cpp:

Canonical link: https://commits.webkit.org/298436@main
